homepage-dashboard: init at 0.6.21

jnsgruk opened the pull request:
###### Description of changes

Introduces a new package for [homepage](https://gethomepage.dev) as well as a NixOS module and some tests.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

jnsgruk committed:
homepage-dashboard: init at 0.6.21

jnsgruk committed:
nixos/homepage-dashboard: init

jnsgruk committed:
tests/homepage-dashboard: add tests for homepage

SuperSandro2000 commented on pkgs/servers/homepage/server.nix:
```diff
@@ -0,0 +1,43 @@
+  npmDepsHash = "sha256-O6SQYx5vxscMsbWv0ynUYqdUkOp/nMtdvlZ/Mp95sBY=";
+
+  # Patch accepted upstream; this can be removed in next release.
+  patches = [ ./config-dir.patch ];
```
Can we fetchpatch then?

jnsgruk replied:
Yep, done!

SuperSandro2000 commented on pkgs/servers/homepage/server.nix:
```diff
@@ -0,0 +1,43 @@
+
+buildNpmPackage rec {
+  pname = "homepage-server";
+  version = "v0.6.21";
+
+  src = fetchFromGitHub {
+    owner = "benphelps";
+    repo = "homepage";
+    rev = "${version}";
+    sha256 = "sha256-kjxA02hJj/GAQ0fM1xTtXAnZSQgVyE+EMRrXis1Vr+o=";
```
```suggestion
  version = "0.6.21";

  src = fetchFromGitHub {
    owner = "benphelps";
    repo = "homepage";
    rev = "v${version}";
    sha256 = "sha256-kjxA02hJj/GAQ0fM1xTtXAnZSQgVyE+EMRrXis1Vr+o=";
```

jnsgruk replied:
Fixed

SuperSandro2000 commented on pkgs/servers/homepage/default.nix:
```diff
@@ -0,0 +1,41 @@
+let
+  homepageServer = callPackage ./server.nix { };
+
+  homepageLauncher = writeShellScriptBin "homepage" ''
+    export HOMEPAGE_CONFIG_DIR="''${HOMEPAGE_CONFIG_DIR:-$XDG_CONFIG_HOME/homepage}"
+    export PORT="''${PORT:-3000}"
+    exec ${nodejs}/bin/node ${homepageServer}/server.js
+  '';
```
This sounds more like a task for makeWrapper

jnsgruk replied:
Done!

SuperSandro2000 commented on pkgs/servers/homepage/default.nix:
```diff
@@ -0,0 +1,41 @@
+    exec ${nodejs}/bin/node ${homepageServer}/server.js
+  '';
+
+  version = homepageServer.version;
+in
+
+symlinkJoin {
+  inherit version;
```
```suggestion
in

symlinkJoin {
  inherit (homepageServer) version;
```

jnsgruk replied:
Removed this and tidied up package a bit

SuperSandro2000 commented on nixos/tests/homepage.nix:
```diff
@@ -0,0 +1,16 @@
+  name = "homepage";
+  meta.maintainers = with lib.maintainers; [ jnsgruk ];
+
+  nodes.machine =
+    { pkgs, ... }:
+    { services.homepage.enable = true; };
```
```suggestion
  nodes.machine = { pkgs, ... }: {
    services.homepage.enable = true;
  };
```

jnsgruk replied:
Fixed

SuperSandro2000 commented on nixos/modules/services/misc/homepage.nix:
```diff
@@ -0,0 +1,47 @@
+        Type = "simple";
+        DynamicUser = true;
+        StateDirectory = "homepage";
+        ExecStart = "${pkgs.homepage}/bin/homepage";
```
```suggestion
        ExecStart = "${lib.getExe pkgs.homepage}";
```

jnsgruk replied:
done!

SuperSandro2000 reviewed (commented):
homepage is a very generic name, do we have some good alternatives?

jnsgruk commented on pkgs/servers/homepage/server.nix:
```diff
@@ -0,0 +1,43 @@
+    cp -r .next/static $out/.next/static
+  '';
+
+  doDist = false;
```
Fixed :)

jnsgruk commented on pkgs/servers/homepage/server.nix:
```diff
@@ -0,0 +1,43 @@
+    python3
+  ];
+
+  PYTHON = "${python3}/bin/python";
```
Thanks :)

jnsgruk commented on pkgs/servers/homepage/server.nix:
```diff
@@ -0,0 +1,43 @@
+
+  buildInputs = [
+    nodePackages.node-gyp-build
+    python3
```
Thanks, fixed.

jnsgruk commented on pkgs/servers/homepage/server.nix:
```diff
@@ -0,0 +1,43 @@
+  patches = [ ./config-dir.patch ];
+
+  postPatch = ''
+    mkdir -p config
```
Thanks, fixed.

jnsgruk commented on pkgs/servers/homepage/default.nix:
```diff
@@ -0,0 +1,41 @@
+
+  passthru = {
+    tests.smoke-test = nixosTests.homepage;
+  };
```
Done!

jnsgruk commented:
@SuperSandro2000 thanks for the detailed feedback. I agree that `homepage` was a bit generic. I've addressed your feedback and changed the name to `homepage-dashboard` - I think that's a bit less ambiguous.

drupol commented on pkgs/servers/homepage-dashboard/default.nix:
```diff
@@ -0,0 +1,76 @@
+, python3
+, lib
+, fetchpatch
+, makeWrapper
```
```suggestion
, makeBinaryWrapper
```

I think it's better to use this to avoid issue on darwin arch.

drupol commented on pkgs/servers/homepage-dashboard/default.nix:
```diff
@@ -0,0 +1,76 @@
+    owner = "benphelps";
+    repo = "homepage";
+    rev = "v${version}";
+    sha256 = "sha256-kjxA02hJj/GAQ0fM1xTtXAnZSQgVyE+EMRrXis1Vr+o=";
```
```suggestion
    hash = "sha256-kjxA02hJj/GAQ0fM1xTtXAnZSQgVyE+EMRrXis1Vr+o=";
```

drupol commented on pkgs/servers/homepage-dashboard/default.nix:
```diff
@@ -0,0 +1,76 @@
+
+  env.PYTHON = "${python3}/bin/python";
+
+  installPhase = ''
```
Missing phase hooks

jnsgruk replied:
Sorry, what do you mean by this?

drupol commented on nixos/modules/services/misc/homepage-dashboard.nix:
```diff
@@ -0,0 +1,53 @@
+{
+  options = {
+    services.homepage-dashboard = {
+      enable = lib.mkEnableOption (lib.mdDoc "Homepage Dashboard");
```
```suggestion
      enable = lib.mkEnableOption (lib.mdDoc "Homepage Dashboard");
      package = lib.mkPackageOptionMD pkgs "homepage-dashboard" { };      
```

drupol commented on nixos/modules/services/misc/homepage-dashboard.nix:
```diff
@@ -0,0 +1,53 @@
+        Type = "simple";
+        DynamicUser = true;
+        StateDirectory = "homepage-dashboard";
+        ExecStart = "${pkgs.homepage-dashboard}/homepage";
```
```suggestion
        ExecStart = "${lib.getExe cfg.package}";
```

drupol commented on pkgs/servers/homepage-dashboard/default.nix:
```diff
@@ -0,0 +1,76 @@
+    inherit (nixosTests) homepage;
+  };
+
+  meta = {
```
add `mainProgram = "homepage";` ?

drupol reviewed (changes requested):
Hi,

I just added a few comments.

drupol commented on pkgs/servers/homepage-dashboard/default.nix:
```diff
@@ -0,0 +1,76 @@
+
+  env.PYTHON = "${python3}/bin/python";
+
+  installPhase = ''
```
```suggestion
  installPhase = ''
    runHook preInstall
```

drupol commented on pkgs/servers/homepage-dashboard/default.nix:
```diff
@@ -0,0 +1,76 @@
+    wrapProgram $out/homepage \
+      --set-default PORT 3000 \
+      --set-default HOMEPAGE_CONFIG_DIR /var/lib/homepage-dashboard
+  '';
```
```suggestion
    runHook postInstall
  '';
```

RaitoBezarius commented on nixos/modules/services/misc/homepage-dashboard.nix:
```diff
@@ -0,0 +1,55 @@
+        PORT = "${toString cfg.listenPort}";
+      };
+
+      serviceConfig = {
```
On the long run, extra hardening would be nice. :)

RaitoBezarius reviewed (approved):
LGTM.

bjornfor commented:
This software advertises itself as a "fully static" dashboard (https://gethomepage.dev/v0.7.3/) that is configurable with yaml, but when I look here I find it runs node as a service and lacks declarative configuration support?

Can we add a NixOS settings option?

bjornfor commented:
> This software advertises itself as a "fully static" dashboard (https://gethomepage.dev/v0.7.3/) that is configurable with yaml, but when I look here I find it runs node as a service and lacks declarative configuration support?
> 
> Can we add a NixOS settings option?

For now I'm doing this:

```nix
...
let
  # Attrset of filenames (/var/lib/private/homepage-dashboard/${attr}) to
  # content.
  #
  # For configuration options and examples, please see:
  # https://gethomepage.dev/latest/configs/.
  settingsRaw = {
    "bookmarks.yaml" = ''
      #- Developer:
      #    - Github:
      #        - abbr: GH2
      #          href: https://github.com/
    '';

    "custom.css" = ''
    '';

    "custom.js" = ''
    '';

    "docker.yaml" = ''
    '';

    "kubernetes.yaml" = ''
    '';

    "services.yaml" = ''
      - local services:
          - My First Service1:
              href: http://localhost/
              description: Homepage is awesome
          - My First Service2:
              href: http://localhost/
              description: Homepage is awesome
          - Gitea:
              href: http://localhost/
              icon: gitea
              description: Git repository hosting, issue tracking and CI server
    '';

    "settings.yaml" = ''
      title: My Awesome Homepage
      background: https://images.unsplash.com/photo-1502790671504-542ad42d5189?auto=format&fit=crop&w=2560&q=80
      #background: https://wallpapercave.com/wp/wp2561067.jpg
    '';

    "widgets.yaml" = ''
      - resources:
          cpu: true
          memory: true
          disk: /

      #- search:
      #    provider: duckduckgo
      #    target: _blank
    '';
  };
in
...
  systemd.tmpfiles.rules =
    lib.mapAttrsToList (n: v: ''L+ /var/lib/private/homepage-dashboard/${n} - - - - ${builtins.toFile n v}'') settingsRaw;
...
```

Seeing how the service instantly reloads changed YAML files, I think we should have *opt-in* declarative config, in case people want imperative configuration (which homepage-dashboard does so well).

jnsgruk commented:
Hey @bjornfor, yeh I'm definitely open to this, but haven't quite gotten around to it. Happy to review any changes to the module if you get there before I do?

jnsgruk commented:
@bjornfor - see #291554 :)
